airbyte-cdk/bulk/core/load/src/testFixtures/kotlin/io/airbyte/cdk/load/test/util/destination_process/DockerizedDestination.kt

```diff
@@ -106,20 +123,89 @@ class DockerizedDestination(
+                        if (message.type == AirbyteMessage.Type.LOG) {
+                            // Don't capture logs, just echo them directly to our own stdout
+                            val combinedMessage =
+                                message.log.message +
+                                    (if (message.log.stackTrace != null)
+                                        (System.lineSeparator() + "Stack Trace: " + message.log.stackTrace)
+                                    else "")
+                            when (message.log.level) {
+                                null, // this should be impossible, treat it as error
+                                AirbyteLogMessage.Level.FATAL, // klogger doesn't have a fatal level
+                                AirbyteLogMessage.Level.ERROR -> logger.error { combinedMessage }
+                                AirbyteLogMessage.Level.WARN -> logger.warn { combinedMessage }
+                                AirbyteLogMessage.Level.INFO -> logger.info { combinedMessage }
+                                AirbyteLogMessage.Level.DEBUG -> logger.debug { combinedMessage }
+                                AirbyteLogMessage.Level.TRACE -> logger.trace { combinedMessage }
+                            }
+                        } else {
+                            destinationOutput.accept(message)
+                        }
+                    }
+                    stdoutDrained.complete(Unit)
+                }
+                launch {
+                    // Consume stderr. Connectors shouldn't really use this,
+                    // and whatever this stream contains, it's almost certainly not valid messages.
+                    // Dump it straight to our own stderr.
+                    process.errorReader().forEachLine { logger.error { it } }
+                    stderrDrained.complete(Unit)
+                }
+            }
+        }
     }
 
     override fun sendMessage(message: AirbyteMessage) {
-        // push a message to the docker process' stdin
-        TODO("Not yet implemented")
+        destinationStdin.write(Jsons.writeValueAsString(message))
+        destinationStdin.newLine()
     }
 
     override fun readMessages(): List<AirbyteMessage> {
-        // read everything from the process' stdout
-        TODO("Not yet implemented")
+        return destinationOutput.newMessages()
     }
 
     override fun shutdown() {
-        // close stdin, wait until process exits
-        TODO("Not yet implemented")
+        destinationStdin.close()
+        // The old cdk had a 1-minute timeout here. That seems... weird?
+        // We can just rely on the junit timeout, presumably?
+        process.waitFor()
+        // Wait for ourselves to drain stdout/stderr. Otherwise we won't capture
+        // all the destination output (logs/trace messages).
+        runBlocking {
+            stdoutDrained.join()
+            stderrDrained.join()
```

Comment:
I didn't even realize that these extend `Job`. That's neat, that means you could cancel them as well if you needed to.